common/djangoapps/entitlements/models.py

```diff
@@ -302,6 +304,45 @@ def get_active_entitlements_for_user(cls, user):
             enrollment_course_run=None
         ).select_related('user').select_related('enrollment_course_run')
 
+    @classmethod
+    def check_for_existing_entitlement_and_enroll(cls, user, course_run_key):
+        """
+        Looks at the User's existing entitlements to see if the user already has a Course Entitlement for the course run
+        provided in the course_key.  If the user does have an Entitlement with no run set, the User is enrolled in the mode
+        set in the Entitlement.
+
+        Note: This method assumes that eligibility to enroll in the course has already been checked.
```

Comment:
Enroll eligibility may have already been checked for the course but are we sure its checking upgradeability and such?